File: Python/Entities/eslEntity2.py
import msgpack
from dataclasses import dataclass, field
from Enums.pattern import Patterns
from Enums.pageIndex import PageIndexes

@dataclass
class ESLEntity2:
    """
    ESL entity for eStation2
    """
    TagID: str = ""
    Pattern: Patterns = Patterns.UpdateDisplay
    PageIndex: PageIndexes = PageIndexes.P0
    R: bool = False
    G: bool = False
    B: bool = False
    Times: int = 0
    Token: int = 0
    CurrentKey: str = ""
    NewKey: str = ""
    Bytes: bytes = field(default_factory=bytes)
    Compress: bool = True

    def to_msgpack(self) -> bytes:
        # Serializes the object to MessagePack format using key order
        data = [
            self.TagID,
            self.Pattern,
            self.PageIndex,
            self.R,
            self.G,
            self.B,
            self.Times,
            self.Token,
            self.CurrentKey,
            self.NewKey,
            self.Bytes,
            self.Compress
        ]
        return msgpack.packb(data)

    @staticmethod
    def from_msgpack(packed: bytes):
        # Deserializes MessagePack data to ESLEntity
        data = msgpack.unpackb(packed)
        return ESLEntity2(
            TagID=data[0],
            Pattern=data[1],
            PageIndex=data[2],
            R=data[3],
            G=data[4],
            B=data[5],
            Times=data[6],
            Token=data[7],
            CurrentKey=data[8],
            NewKey=data[9],
            Bytes=data[10],
            Compress=data[11]
        )

File: Java/src/main/java/com/estation/demo/entities/EslEntity2.java
package com.estation.demo.entities;

import com.estation.demo.enums.PageIndex;
import com.estation.demo.enums.Pattern;
import com.estation.demo.util.MessagePackUtil;
import java.util.List;

public record EslEntity2(
        String tagId,
        Pattern pattern,
        PageIndex pageIndex,
        boolean r,
        boolean g,
        boolean b,
        int times,
        int token,
        String currentKey,
        String newKey,
        byte[] bytes,
        boolean compress
) {
    public List<Object> toPayloadList() {
        return List.of(
                tagId,
                pattern.value(),
                pageIndex.value(),
                r,
                g,
                b,
                times,
                token,
                currentKey,
                newKey,
                bytes,
                compress
        );
    }

    public byte[] toMessagePack() {
        return MessagePackUtil.pack(toPayloadList());
    }

    public static EslEntity2 fromMessagePack(byte[] data) {
        List<Object> values = MessagePackUtil.asList(MessagePackUtil.unpack(data));
        return new EslEntity2(
                MessagePackUtil.asString(values.get(0)),
                Pattern.fromValue(MessagePackUtil.asInt(values.get(1))),
                PageIndex.fromValue(MessagePackUtil.asInt(values.get(2))),
                MessagePackUtil.asBoolean(values.get(3)),
                MessagePackUtil.asBoolean(values.get(4)),
                MessagePackUtil.asBoolean(values.get(5)),
                MessagePackUtil.asInt(values.get(6)),
                MessagePackUtil.asInt(values.get(7)),
                MessagePackUtil.asString(values.get(8)),
                MessagePackUtil.asString(values.get(9)),
                MessagePackUtil.asBinary(values.get(10)),
                MessagePackUtil.asBoolean(values.get(11))
        );
    }
}
